query-algebrizer/src/validate.rs

```diff
@@ -229,4 +274,177 @@ mod tests {
+        let query = r#"[:find ?artist
+                        :where [?artist :artist/name]
+                               (not-join [?artist]
+                                   [?release :release/artists ?artist]
+                                   [?release :release/year 1970])]"#;
+        let parsed = parse_find_string(query).expect("expected successful parse");
+        let clauses = valid_not_join(parsed, UnifyVars::Explicit(vec![Variable::from_valid_name("?artist")]));
+
+        // // Let's do some detailed parse checks.
+        let mut parts = clauses.into_iter();
+        match (parts.next(), parts.next(), parts.next()) {
+            (Some(clause1), Some(clause2), None) => {
+                assert_eq!(
+                    clause1,
+                    WhereNotClause::Clause(
+                        WhereClause::Pattern(Pattern {
+                            source: None,
+                            entity: PatternNonValuePlace::Variable(Variable::from_valid_name("?release")),
+                            attribute: ident("release", "artists"),
+                            value: PatternValuePlace::Variable(Variable::from_valid_name("?artist")),
+                            tx: PatternNonValuePlace::Placeholder,
+                        })));
+                assert_eq!(
+                    clause2,
+                    WhereNotClause::Clause(
+                        WhereClause::Pattern(Pattern {
+                            source: None,
+                            entity: PatternNonValuePlace::Variable(Variable::from_valid_name("?release")),
+                            attribute: ident("release", "year"),
+                            value: PatternValuePlace::EntidOrInteger(1970),
+                            tx: PatternNonValuePlace::Placeholder,
+                        })));
+            },
+            _ => panic!(),
+        };
+    }
+    
+    /// Test that a `not-join` that does not use the joining var fails to validate.
+    #[test]
+    fn test_invalid_explicit_not_join_non_matching_join_vars() {
+        let query = r#"[:find ?artist
+                        :where [?artist :artist/name]
+                               (not-join [?artist]
+                                   [?release :release/artists "Pink Floyd"]
+                                   [?release :release/year 1970])]"#;
+        let parsed = parse_find_string(query).expect("expected successful parse");
+        let mut nots = parsed.where_clauses.iter().filter(|&x| match *x {
+            WhereClause::NotJoin(_) => true,
+            _ => false,
+        });
+
+        let clause = nots.next().unwrap().clone();
+        assert_eq!(None, nots.next());
+
+        match clause {
+            WhereClause::NotJoin(not_join) => assert!(validate_not_join(&not_join).is_err()),
+            _ => panic!(),
+        }
+    }    
+    
+    /// Test that a `not-join` that does not match any vars declared inside it fails to validate.
```

Comment:
This query is valid. Again, it's almost certain not to be a meaningful query — the variable named `?year` will be bound to releases, and won't join against the rest of the clauses — but it's valid. It means "Find all artists, but only artists with no releases unless anybody released a record in 1970".